### DISCUSSION TITLE
Can not use localhost with Callbacks

### DISCUSSION THREAD

**Discussion by PhillipOtey** [fastapi/fastapi#6926] [Questions] [answered]
I can't use localhost with Callbacks. It's something to do with it's trying to find the TLD of the address. The code works fine with external websites but doesn't work with other local hosts or with virtual machines. For reference I am able to use this with requests and grequests just fine.
For instance `https://www.Google.com` works,
However `http://localhost:3000/` doesn't and gives the error below in `/docs`:

###Example
I have used exactly what is in the example for Callback. 
https://fastapi.tiangolo.com/advanced/openapi-callbacks/?h=+call
Callback url: "http://localhost:3000/"
```
{
    "detail": [
        {
            "Loc": [
                "querry",
                "Callbacl_url"
            ],
            "msg": "URL host invalid, top level domain required"
            "type": "value_error.url.host"
        }
    ]
}
```
‐----------------------

Here is the python code from said example for the open api. I use `uvicorn` to run the app you could use the next two shell scripts to get two servers running.

```Sh script (in bash terminal)
uvicorn my_app.main:app --reload --port 4000
```
```Sh script (in bash terminal)
uvicorn my_app.main:app --reload --port 3000
```

```Python
#my_app
from typing import Optional

from fastapi import APIRouter, FastAPI
from pydantic import BaseModel, HttpUrl

app = FastAPI()


class Invoice(BaseModel):
    id: str
    title: Optional[str] = None
    customer: str
    total: float


class InvoiceEvent(BaseModel):
    description: str
    paid: bool


class InvoiceEventReceived(BaseModel):
    ok: bool


invoices_callback_router = APIRouter()


@invoices_callback_router.post(
    "{$callback_url}/invoices/{$request.body.id}", response_model=InvoiceEventReceived
)
def invoice_notification(body: InvoiceEvent):
    pass


@app.post("/invoices/", callbacks=invoices_callback_router.routes)
def create_invoice(invoice: Invoice, callback_url: Optional[HttpUrl] = None):
    """
    Create an invoice.

    This will (let's imagine) let the API user (some external developer) create an
    invoice.

    And this path operation will:

    * Send the invoice to the client.
    * Collect the money from the client.
    * Send a notification back to the API user (the external developer), as a callback.
        * At this point is that the API will somehow send a POST request to the
            external API with the notification of the invoice event
            (e.g. "payment successful").
    """
    # Send the invoice, collect the money, send the notification (the callback)
    return {"msg": "Invoice received"}

@app.get("/")
def read_root():
    return {"Hello": "World"}
```

### Description
I can't use Callbacks with localhost because of a parse search for the TLD (top level domain) or something causing the error message above.

### Environment

* OS: [Windows]:
* FastAPI Version [e.g. 0.63.0]:

* Python version:  3.7.1

```bash version 4.4.23
```

**Comment by Mause** [comment, marked as answer]
Use `AnyHttpUrl` instead of `HttpUrl` - docs are here: https://pydantic-docs.helpmanual.io/usage/types/#urls

**Comment by tiangolo** [comment]
Thanks for the help here @Mause ! :clap: :bow: 

Thanks for reporting back and closing the issue @PhillipOtey  :+1:
